lib/client/index.js

```diff
@@ -0,0 +1,1710 @@
+      nowDate = now;
+
+      updateCurrentSGV(latestSGV);
+      updateClockDisplay();
+      updateTimeAgo();
+      updatePlugins(nowData, nowDate);
+
+    }
+
+    xScale.domain(brush.extent());
+
+    // get slice of data so that concatenation of predictions do not interfere with subsequent updates
+    var focusData = data.slice();
+
+    if (sbx.pluginBase.forecastPoints) {
+      focusData = focusData.concat(sbx.pluginBase.forecastPoints);
+    }
+
+    // bind up the focus chart data to an array of circles
+    // selects all our data into data and uses date function to get current max date
+    var focusCircles = focus.selectAll('circle').data(focusData, dateFn);
+
+    var focusRangeAdjustment = foucusRangeMS === THREE_HOURS_MS ? 1 : 1 + ((foucusRangeMS - THREE_HOURS_MS) / THREE_HOURS_MS / 8);
+
+    var dotRadius = function(type) {
+      var radius = prevChartWidth > WIDTH_BIG_DOTS ? 4 : (prevChartWidth < WIDTH_SMALL_DOTS ? 2 : 3);
+      if (type === 'mbg') {
+        radius *= 2;
+      } else if (type === 'forecast') {
+        radius = Math.min(3, radius - 1);
+      } else if (type === 'rawbg') {
+        radius = Math.min(2, radius - 1);
+      }
+
+      return radius / focusRangeAdjustment;
+    };
+
+    function isDexcom(device) {
+      return device && device.toLowerCase().indexOf('dexcom') === 0;
+    }
+
+    function prepareFocusCircles(sel) {
+      var badData = [];
+      sel.attr('cx', function (d) {
+        if (!d) {
+          console.error('Bad data', d);
+          return xScale(new Date(0));
+        } else if (!d.mills) {
+          console.error('Bad data, no mills', d);
+          return xScale(new Date(0));
+        } else {
+          return xScale(new Date(d.mills));
+        }
+      })
+        .attr('cy', function (d) {
+          var scaled = sbx.scaleEntry(d);
+          if (isNaN(scaled)) {
+            badData.push(d);
+            return yScale(scaleBg(450));
+          } else {
+            return yScale(scaled);
+          }
+        })
+        .attr('fill', function (d) { return d.type === 'forecast' ? 'none' : d.color; })
+        .attr('opacity', function (d) { return futureOpacity(d.mills - latestSGV.mills); })
+        .attr('stroke-width', function (d) { return d.type === 'mbg' ? 2 : d.type === 'forecast' ? 1 : 0; })
+        .attr('stroke', function (d) {
+          return (isDexcom(d.device) ? 'white' : d.type === 'forecast' ? d.color : '#0099ff');
+        })
+        .attr('r', function (d) { return dotRadius(d.type); });
+
+      if (badData.length > 0) {
+        console.warn('Bad Data: isNaN(sgv)', badData);
+      }
+
+      return sel;
+    }
+
+    // if already existing then transition each circle to its new position
+    prepareFocusCircles(focusCircles.transition().duration(UPDATE_TRANS_MS));
+
+    // if new circle then just display
+    prepareFocusCircles(focusCircles.enter().append('circle'))
+      .on('mouseover', function (d) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l510)